From: Rene Kita <mail@rkta.de>
Date: Tue Sep 17 15:32:20 CEST 2024
Subject: Replace STR_SIZE_MAX with STR_LEN_MAX
Patch: patches/Replace-STR_SIZE_MAX-with-STR_LEN_MAX.patch

We are usually dealing with string length and not with the size a.k.a.
the length plus the terminating nul character or the allocated
area_size. The area_size is an implementation detail and should be as
little visible to the user as possible.

Use a STR_LEN_MAX to avoid dealing with of by errors.
---
 Str.c  |   17 +++++++++--------
 Str.h  |    4 ++--
 main.c |    2 +-
 3 files changed, 12 insertions(+), 11 deletions(-)

--- a/Str.c
+++ b/Str.c
@@ -25,6 +25,7 @@
 #include "myctype.h"
 
 #define INITIAL_STR_SIZE 32
+#define STR_SIZE_MAX (STR_LEN_MAX + 1)
 
 static Str
 Strnulterm(Str x)
@@ -54,8 +55,8 @@ Strnew_size(int n)
     Str x = GC_MALLOC(sizeof(struct _Str));
     if (x == NULL)
 	exit(1);
-    if (n < 0 || n >= STR_SIZE_MAX)
-	n = STR_SIZE_MAX - 1;
+    if (n < 0 || n > STR_LEN_MAX)
+	n = STR_LEN_MAX;
     else if (n + 1 < INITIAL_STR_SIZE)
 	n = INITIAL_STR_SIZE - 1;
     x->ptr = GC_MALLOC_ATOMIC(n + 1);
@@ -182,8 +183,8 @@ Strcopy_charp(Str x, const char *y)
 	return;
     }
     len = strlen(y);
-    if (len < 0 || len >= STR_SIZE_MAX)
-	len = STR_SIZE_MAX - 1;
+    if (len < 0 || len > STR_LEN_MAX)
+	len = STR_LEN_MAX;
     if (x->area_size < len + 1) {
 	x->ptr = GC_REALLOC(x->ptr, len + 1);
 	if (x->ptr == NULL)
@@ -205,8 +206,8 @@ Strcopy_charp_n(Str x, const char *y, in
 	Strnulterm(x);
 	return;
     }
-    if (len < 0 || len >= STR_SIZE_MAX)
-	len = STR_SIZE_MAX - 1;
+    if (len < 0 || len > STR_LEN_MAX)
+	len = STR_LEN_MAX;
     if (x->area_size < len + 1) {
 	x->ptr = GC_REALLOC(x->ptr, len + 1);
 	if (x->ptr == NULL)
@@ -226,7 +227,7 @@ Strcat_charp_n(Str x, const char *y, int
     if (y == NULL || n == 0)
 	return;
     if (n < 0)
-	n = STR_SIZE_MAX - 1;
+	n = STR_LEN_MAX;
     newlen = x->length + n + 1;
     if (newlen <= 0 || newlen > STR_SIZE_MAX) {
 	newlen = STR_SIZE_MAX;
@@ -371,7 +372,7 @@ Strdelete(Str s, int pos, int n)
     if (pos < 0 || s->length < pos)
 	return;
     if (n < 0)
-	n = STR_SIZE_MAX - pos - 1;
+	n = STR_LEN_MAX - pos;
     if (s->length <= pos + n) {
 	s->length = pos;
 	Strnulterm(s);
--- a/Str.h
+++ b/Str.h
@@ -66,8 +66,8 @@ Str Strfgetall(FILE *);
 
 void Strgrow(Str s);
 
-#define STR_SIZE_MAX (INT_MAX / 32)
-#define Strcat_char(x,y) (((x)->length+1>=STR_SIZE_MAX)?0:(((x)->length+1>=(x)->area_size)?Strgrow(x),0:0,(x)->ptr[(x)->length++]=(y),(x)->ptr[(x)->length]=0))
+#define STR_LEN_MAX (INT_MAX / 32 - 1)
+#define Strcat_char(x,y) (((x)->length>=STR_LEN_MAX)?0:(((x)->length+1>=(x)->area_size)?Strgrow(x),0:0,(x)->ptr[(x)->length++]=(y),(x)->ptr[(x)->length]=0))
 #define Strcmp(x,y)                  strcmp((x)->ptr,(y)->ptr)
 #define Strcmp_charp(x,y)            strcmp((x)->ptr,(y))
 #define Strcasecmp(x,y)              strcasecmp((x)->ptr,(y)->ptr)
--- a/main.c
+++ b/main.c
@@ -480,7 +480,7 @@ main(int argc, char **argv)
 	    /* Don't use hostname if it is truncated.  */
 	    hostname[HOST_NAME_MAX + 1] = '\0';
 	    hostname_len = strlen(hostname);
-	    if (hostname_len <= HOST_NAME_MAX && hostname_len < STR_SIZE_MAX)
+	    if (hostname_len <= HOST_NAME_MAX && hostname_len <= STR_LEN_MAX)
 		HostName = allocStr(hostname, hostname_len);
 	}
     }
